feat: add ed25519 support to JWK (public keys)

edudobay opened the pull request:
Ed25519 key support has been added in #343 for the "standard" key format, but not yet for JWKs. (I'm opening a new PR, as I had previously opened #414 against a different base branch.)

Reference documentation: [RFC8037](https://datatracker.ietf.org/doc/html/rfc8037)

bshaffer commented on tests/JWKTest.php:
```diff
@@ -151,6 +151,7 @@ public function provideDecodeByJwkKeySet()
         return [
             ['rsa1-private.pem', 'rsa-jwkset.json', 'RS256'],
             ['ecdsa256-private.pem', 'ec-jwkset.json', 'ES256'],
+            ['ed25519-1.sec', 'ed25519-jwkset.json', 'EdDSA'],
```
It looks like you forgot to include these files!

bshaffer replied:
my mistake, we already have these files for other EC tests!

bshaffer commented on src/JWT.php:
```diff
@@ -385,13 +385,26 @@ public static function jsonEncode(array $input): string
+     * @return string A Base64 encoded string with standard characters (+/) and padding (=), when
+     * needed.
+     */
+    public static function urlsafeToStandardB64(string $input): string
```
This naming is confusing to me. For key type OKP, the [spec says](https://www.rfc-editor.org/rfc/rfc8037.html)

> The parameter "x" MUST be present and contain the public key
      encoded using the base64url [[RFC4648](https://www.rfc-editor.org/rfc/rfc4648)] encoding.

This links to _another_ [spec](https://www.rfc-editor.org/rfc/rfc4648#section-5) which says

> This encoding may be referred to as "base64url".  This encoding
   should not be regarded as the same as the "base64" encoding and
   should not be referred to as only "base64"

I think it would be better to call this `base64UrlDecode`

edudobay replied:
I've renamed the function to `convertBase64urlToBase64`. Do you think it's clearer?

It doesn't actually decode base64 data, it just translates from `base64url` to `base64` so that the PHP `base64_decode` function can be used.

bshaffer commented on src/JWK.php:
```diff
@@ -96,11 +101,17 @@ public static function parseKey(array $jwk, string $defaultAlg = null): ?Key
         }
 
-        if (!isset($jwk['alg'])) {
+        $ktyNotRequiringAlg = [
+            // In Octet Key Pair (OKP) keys, the signing algorithm (alg) will not be read from the
+            // JWK, as it can be inferred directly from the curve type (crv).
+            // @see https://datatracker.ietf.org/doc/html/rfc8037#section-3.1
+            'OKP',
+        ];
+        if (!isset($jwk['alg']) && !\in_array($jwk['kty'], $ktyNotRequiringAlg, true)) {
```
I am unfamiliar with OKPs, but by looking at [this example](https://curity.io/resources/learn/jwt-signatures/), those values contain an "alg" of `EdDSA` instead of using the clunky mapping. I'd prefer to keep the algorithm explicit.

edudobay replied:
I've reverted to the original version where the algorithm is always required. I see that the library opted for always requiring an explicit algorithm.

bshaffer commented on src/JWK.php:
```diff
@@ -144,8 +155,28 @@ public static function parseKey(array $jwk, string $defaultAlg = null): ?Key
+                    throw new UnexpectedValueException('Key data must be for a public key');
+                }
+
+                if (! isset($jwk['crv'])) {
```
```suggestion
                if (!isset($jwk['crv'])) {
```

bshaffer commented on src/JWK.php:
```diff
@@ -30,6 +30,11 @@ class JWK
     ];
 
+    // 'crv' identifier => JWT 'alg'
+    private const OKP_CURVES = [
```
This is strange to me, as the values here aren't curves, but algorithms. I am not convinced this is the correct behavior.

edudobay replied:
I've rephrased the concepts to call these "key subtypes", as section 2 of RFC 8037 calls it.

bshaffer commented on src/JWK.php:
```diff
@@ -144,8 +155,28 @@ public static function parseKey(array $jwk, string $defaultAlg = null): ?Key
+                    throw new UnexpectedValueException('x not set');
+                }
+
+                $publicKey = JWT::urlsafeToStandardB64($jwk['x']);
+                $alg = self::OKP_CURVES[$jwk['crv']];
```
I am confused because there is no action taken as far as implementing [decoding a key with ed25519](https://www.rfc-editor.org/rfc/rfc8032#section-5.1). The `JWK` class is `base64url`-decoding the key, and (I am assuming) relying on the to default behavior of `sodium_crypto` to accept this as `ed25519`. There are other possible values here, however, so I think at the very least this needs to throw an exception if `crv` is not `ed25519`.

edudobay replied:
Yes, the `sodium_crypto` functions that are used by the library work specifically with ed25519.

If an unsupported `crv` is provided, an exception will be thrown a few lines above with the message `Unrecognised or unsupported OKP key subtype`.

bshaffer reviewed (commented):
Thank you for submitting this! I have some technical questions, as well as some issues with the implementation here. Hopefully we can get this feature merged!

edudobay commented:
Hi! Sorry for the delay, I've addressed the issues you commented.

I was wondering if it would be useful to throw an exception if the provided `alg` is not supported for the given `kty` (for example, `alg=RS256` and `kty=OKP`). But this verification does not yet exist for the other key types, so I guess this would be the scope of a future PR.

In that case, the mismatch would fall down to the crypto function layer — it would either give an error because the given key is invalid for the given algorithm, or encrypt with the wrong algorithm if by chance the same key is valid for different algorithms.

bshaffer commented on src/JWT.php:
```diff
@@ -399,13 +399,28 @@ public static function jsonEncode(array $input): string
+     *
+     * @see https://www.rfc-editor.org/rfc/rfc4648
+     */
+    public static function convertBase64urlToBase64(string $input): string
```
This is a nit, but to be more consistent with our other namings, we should capitalize `Url` here:
```suggestion
    public static function convertBase64UrlToBase64(string $input): string
```

bshaffer reviewed (commented):
This looks great! Just one minor change and we can merge this!

Thanks again for your contribution.

edudobay commented on tests/data/ed25519-jwkset.json:
```diff
@@ -0,0 +1,10 @@
+            "kid": "jwk1",
+            "kty": "OKP",
+            "crv": "Ed25519",
+            "x": "uOSJMhbKSG4V5xUHS7B9YHmVg_1yVd-G-Io6oBFhSfY"
```
Added the parameter.

edudobay commented:
@bshaffer I've addressed the final comments, so I think this is ready now :)
